File: basketball-csharp/BasketballTickets/Repository/Impl/TicketRepositoryImpl.cs
using BasketballTickets.Domain;
using BasketballTickets.Repository.Abstracts;
using BasketballTickets.Repository.Interfaces;
using BasketballTickets.Utils;
using Microsoft.Data.Sqlite;

namespace BasketballTickets.Repository.Impl;

public class TicketRepositoryImpl : AbstractDBRepository<long, Ticket>, ITicketRepository
{
    private const string JoinSql = """
        SELECT t.id, t.numberOfSeats,
               c.id AS cId, c.name AS cName, c.address AS cAddress,
               m.id AS mId, m.name AS mName, m.ticketPrice, m.totalSeats, m.availableSeats
        FROM tickets t
        JOIN customers c ON t.customerId = c.id
        JOIN matches   m ON t.matchId   = m.id
        """;

    public TicketRepositoryImpl(DatabaseConfig dbConfig) : base(dbConfig) { }

    protected override string GetTableName() => "tickets";

    protected override string GetInsertSql() =>
        "INSERT INTO tickets (customerId, matchId, numberOfSeats) VALUES (@customerId, @matchId, @seats)";

    protected override string GetUpdateSql() =>
        "UPDATE tickets SET customerId=@customerId, matchId=@matchId, numberOfSeats=@seats WHERE id=@id";

    // Tickets au nevoie de JOIN — suprascriem FindById si FindAll
    public override Ticket? FindById(long id)
    {
        Logger.Debug($"FindById({id})");
        using var conn = DbConfig.GetConnection();
        using var cmd = new SqliteCommand(JoinSql + " WHERE t.id=@id", conn);
        cmd.Parameters.AddWithValue("@id", id);
        using var reader = cmd.ExecuteReader();
        return reader.Read() ? ExtractEntity(reader) : null;
    }

    public override IList<Ticket> FindAll()
    {
        Logger.Debug("FindAll()");
        return RunJoinQuery(JoinSql, []);
    }

    protected override Ticket ExtractEntity(SqliteDataReader r)
    {
        var customer = new Customer(
            r.GetInt64(r.GetOrdinal("cId")),
            r.GetString(r.GetOrdinal("cName")),
            r.GetString(r.GetOrdinal("cAddress")));

        var match = new Match(
            r.GetInt64(r.GetOrdinal("mId")),
            r.GetString(r.GetOrdinal("mName")),
            r.GetDouble(r.GetOrdinal("ticketPrice")),
            r.GetInt32(r.GetOrdinal("totalSeats")),
            r.GetInt32(r.GetOrdinal("availableSeats")));

        return new Ticket(
            r.GetInt64(r.GetOrdinal("id")),
            customer,
            match,
            r.GetInt32(r.GetOrdinal("numberOfSeats")));
    }

    protected override void BindInsert(SqliteCommand cmd, Ticket t)
    {
        cmd.Parameters.AddWithValue("@customerId", t.Customer.Id);
        cmd.Parameters.AddWithValue("@matchId", t.Match.Id);
        cmd.Parameters.AddWithValue("@seats", t.NumberOfSeats);
    }

    protected override void BindUpdate(SqliteCommand cmd, Ticket t)
    {
        cmd.Parameters.AddWithValue("@customerId", t.Customer.Id);
        cmd.Parameters.AddWithValue("@matchId", t.Match.Id);
        cmd.Parameters.AddWithValue("@seats", t.NumberOfSeats);
        cmd.Parameters.AddWithValue("@id", t.Id);
    }

    protected override void SetGeneratedId(Ticket entity, long id) => entity.Id = id;

    // ---- Metode specifice domeniului ----

    public IList<Ticket> FindByCustomerNameOrAddress(string? name, string? address)
    {
        Logger.Debug($"FindByCustomerNameOrAddress(name='{name}', address='{address}')");

        var sql = JoinSql + " WHERE 1=1";
        var parameters = new List<(string, object)>();

        if (!string.IsNullOrWhiteSpace(name))
        {
            sql += " AND LOWER(c.name) = LOWER(@name)";
            parameters.Add(("@name", name));
        }
        if (!string.IsNullOrWhiteSpace(address))
        {
            sql += " AND LOWER(c.address) LIKE LOWER(@address)";
            parameters.Add(("@address", $"%{address}%"));
        }

        return RunJoinQuery(sql, parameters);
    }

    public IList<Ticket> FindByMatchId(long matchId)
    {
        Logger.Debug($"FindByMatchId({matchId})");
        return RunJoinQuery(JoinSql + " WHERE t.matchId=@p", [("@p", (object)matchId)]);
    }

    public IList<Ticket> FindByCustomerId(long customerId)
    {
        Logger.Debug($"FindByCustomerId({customerId})");
        return RunJoinQuery(JoinSql + " WHERE t.customerId=@p", [("@p", (object)customerId)]);
    }

    // ---- Helper ----
    private List<Ticket> RunJoinQuery(string sql, IList<(string, object)> parameters)
    {
        var results = new List<Ticket>();
        using var conn = DbConfig.GetConnection();
        using var cmd = new SqliteCommand(sql, conn);
        foreach (var (paramName, value) in parameters)
            cmd.Parameters.AddWithValue(paramName, value);
        using var reader = cmd.ExecuteReader();
        while (reader.Read()) results.Add(ExtractEntity(reader));
        Logger.Debug($"RunJoinQuery → {results.Count} tickets");
        return results;
    }
}

File: java/src/main/java/org/example/repository/impl/TicketRepositoryImpl.java
package org.example.repository.impl;

import org.example.domain.Customer;
import org.example.domain.Match;
import org.example.domain.Ticket;
import org.example.repository.abstracts.AbstractDBRepository;
import org.example.repository.interfaces.TicketRepository;
import org.example.utils.DatabaseConfig;


import java.sql.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class TicketRepositoryImpl
        extends AbstractDBRepository<Long, Ticket>
        implements TicketRepository {

    public TicketRepositoryImpl(DatabaseConfig dbConfig) {
        super(dbConfig);
    }

    @Override
    protected String getTableName() { return "tickets"; }

    @Override
    protected String getInsertSql() {
        return "INSERT INTO tickets (customerId, matchId, numberOfSeats) VALUES (?, ?, ?)";
    }

    @Override
    protected String getUpdateSql() {
        return "UPDATE tickets SET customerId=?, matchId=?, numberOfSeats=? WHERE id=?";
    }

    // Tickets need JOINs — override findById and findAll to use the JOIN query
    private static final String JOIN_SQL = """
            SELECT t.id, t.numberOfSeats,
                   c.id AS cId, c.name AS cName, c.address AS cAddress,
                   m.id AS mId, m.name AS mName, m.ticketPrice, m.totalSeats, m.availableSeats
            FROM tickets t
            JOIN customers c ON t.customerId = c.id
            JOIN matches   m ON t.matchId   = m.id
            """;

    @Override
    public Optional<Ticket> findById(Long id) {
        logger.debug("findById({})", id);
        try (Connection conn = dbConfig.getConnection();
             PreparedStatement ps = conn.prepareStatement(JOIN_SQL + " WHERE t.id=?")) {
            ps.setLong(1, id);
            ResultSet rs = ps.executeQuery();
            if (rs.next()) return Optional.of(extractEntity(rs));
        } catch (SQLException e) {
            logger.error("findById({}) failed: {}", id, e.getMessage(), e);
        }
        return Optional.empty();
    }

    @Override
    public List<Ticket> findAll() {
        logger.debug("findAll()");
        return runJoinQuery(JOIN_SQL, List.of());
    }

    @Override
    protected Ticket extractEntity(ResultSet rs) throws SQLException {
        Customer customer = new Customer(
                rs.getLong("cId"), rs.getString("cName"), rs.getString("cAddress"));
        Match match = new Match(
                rs.getLong("mId"), rs.getString("mName"),
                rs.getDouble("ticketPrice"), rs.getInt("totalSeats"), rs.getInt("availableSeats"));
        return new Ticket(rs.getLong("id"), customer, match, rs.getInt("numberOfSeats"));
    }

    @Override
    protected void bindInsert(PreparedStatement ps, Ticket t) throws SQLException {
        ps.setLong(1, t.getCustomer().getId());
        ps.setLong(2, t.getMatch().getId());
        ps.setInt(3, t.getNumberOfSeats());
    }

    @Override
    protected void bindUpdate(PreparedStatement ps, Ticket t) throws SQLException {
        ps.setLong(1, t.getCustomer().getId());
        ps.setLong(2, t.getMatch().getId());
        ps.setInt(3, t.getNumberOfSeats());
        ps.setLong(4, t.getId());
    }

    @Override
    protected void setGeneratedId(Ticket entity, long id) { entity.setId(id); }

    // ---- Domain-specific queries ----

    @Override
    public List<Ticket> findByCustomerNameOrAddress(String name, String address) {
        logger.debug("findByCustomerNameOrAddress(name='{}', address='{}')", name, address);
        StringBuilder sql = new StringBuilder(JOIN_SQL + " WHERE 1=1");
        List<Object> params = new ArrayList<>();

        if (name != null && !name.isBlank()) {
            sql.append(" AND LOWER(c.name) = LOWER(?)");
            params.add(name);
        }
        if (address != null && !address.isBlank()) {
            sql.append(" AND LOWER(c.address) LIKE LOWER(?)");
            params.add("%" + address + "%");
        }
        return runJoinQuery(sql.toString(), params);
    }

    @Override
    public List<Ticket> findByMatchId(Long matchId) {
        logger.debug("findByMatchId({})", matchId);
        return runJoinQuery(JOIN_SQL + " WHERE t.matchId=?", List.of(matchId));
    }

    @Override
    public List<Ticket> findByCustomerId(Long customerId) {
        logger.debug("findByCustomerId({})", customerId);
        return runJoinQuery(JOIN_SQL + " WHERE t.customerId=?", List.of(customerId));
    }

    // ---- Helper ----
    private List<Ticket> runJoinQuery(String sql, List<Object> params) {
        List<Ticket> result = new ArrayList<>();
        try (Connection conn = dbConfig.getConnection();
             PreparedStatement ps = conn.prepareStatement(sql)) {
            for (int i = 0; i < params.size(); i++) ps.setObject(i + 1, params.get(i));
            ResultSet rs = ps.executeQuery();
            while (rs.next()) result.add(extractEntity(rs));
        } catch (SQLException e) {
            logger.error("runJoinQuery failed: {}", e.getMessage(), e);
        }
        logger.debug("runJoinQuery → {} tickets", result.size());
        return result;
    }
}

